Cloud archiving

merrygoat opened the pull request:
Fixes #9 (at last)

merrygoat committed:
Implement skeleton of DropboxCloudProvider

merrygoat committed:
Separate CloudProvider and CloudProviderType

merrygoat committed:
Rearrange functions to split base CloudProvider

Split base class from its implementations.
Add mock tests for basic Dropbox functions.

merrygoat committed:
Refactor file upload method

merrygoat committed:
Separate file read method

merrygoat committed:
Add tests for normalise path

merrygoat committed:
Add tests for path normalisation, file upload and file reading

merrygoat committed:
Extract get client modified method

merrygoat committed:
Refactor archive_file method

merrygoat committed:
Separate get_token from DropboxCloudProvider init

merrygoat committed:
Separate file exclusion method

merrygoat committed:
Combine archive directory methods

merrygoat committed:
Fix wrong dropbox upload directory

merrygoat committed:
Move some general functions to cloud.py

merrygoat committed:
Add directory exclusion back in

merrygoat committed:
Add tests for get_token

merrygoat committed:
Remove get config from dropbox tests

merrygoat committed:
Add session upload type for large files

merrygoat committed:
Add FileToUpload object

merrygoat committed:
Add progress bar to session upload

aplowman committed:
Fix some bugs

aplowman committed:
Fix bug with upload remote path

aplowman committed:
Merge branch 'dev' into cloud_archiving

aplowman commented on hpcflow/archive/cloud/dropbox_cp.py:
```diff
@@ -0,0 +1,369 @@
+        FileMetadata of file if file uploaded. If file not uploaded return None.
+        """
+        # Get the last modified time of the file on Dropbox
+        dropbox_modified = self._get_dropbox_file_modified_time(file.dropbox_path)
+        if not dropbox_modified:
+            # File does not exist on dropbox
+            return self._choose_session_type(file)
+        if file.last_modified == dropbox_modified:
+            # If file on Dropbox has the same modified time as the client file then don't upload.
+            return None
+        elif file.last_modified > dropbox_modified:
+            # If the client file is newer than the one on dropbox then overwrite the one on dropbox
+            file.overwrite_mode = dropbox.dropbox.files.WriteMode('overwrite')
+
+        return self._choose_session_type(file)
+
+    def _choose_session_type(self, file: FileToUpload) -> dropbox.files.FileMetadata:
+        """Choose whether to upload the file as a simple upload or a session upload
+        depending on its size.
+
+        Parameters
+        ----------
+        file
+            `FileToUpload` object describing file to be uploaded
+        """
+
+        try:
+            if file.size <= CHUNK_SIZE:
+                return self.simple_upload(file)
+            else:
+                return self.session_upload(file)
+        except Exception as err:
+            raise CloudProviderError(f'Cloud provider error: {err}')
```
This is a little confusing, since the method `self._choose_session_type` actually uploads the `file`, which isn't obvious from its name. I would prefer if we remove the `_choose_session_type` method and place its logic inside `_archive_file`:

```suggestion
        dropbox_modified = self._get_dropbox_file_modified_time(file.dropbox_path)
        do_upload = False
        if not dropbox_modified:
            # File does not exist on dropbox
            do_upload = True
        elif file.last_modified > dropbox_modified:
            # If the client file is newer than the one on dropbox then overwrite the one on dropbox
            file.overwrite_mode = dropbox.dropbox.files.WriteMode('overwrite')
            do_upload = True            

        if do_upload:
            try:
                if file.size <= CHUNK_SIZE:
                    return self.simple_upload(file)
                else:
                    return self.session_upload(file)
            except Exception as err:
                raise CloudProviderError(f'Cloud provider error: {err}')        

```

merrygoat replied:
For now I have just renamed the function. I think extracting logic to the superclass is easier done when a second provider is added. FileToUpload can move up a level too probably.

aplowman commented on hpcflow/archive/cloud/dropbox_cp.py:
```diff
@@ -0,0 +1,369 @@
+        file
+            `FileToUpload` object describing file to be uploaded
+        """
+        progress = tqdm.tqdm(total=file.size, desc=f"Uploading file: {file.path.name}",
```
This is a nice idea, but this method will be called non-interactively 99% of the time; so I am not sure how this would work when outputting to `stdout`?

aplowman commented on hpcflow/archive/cloud/dropbox_cp.py:
```diff
@@ -0,0 +1,369 @@
+import os
```
Could we add this module (and the later `onedrive` module) back into a `providers` sub-package? The class hierachy is then clear without opening the files.

aplowman commented on hpcflow/archive/archive.py:
```diff
@@ -70,7 +78,8 @@ class Archive(Base):
     host = Column(String(255))
-    cloud_provider = Column(Enum(CloudProvider))
+    cloud_provider_type = Column(Enum(CloudProviderType))
+    cloud_provider = Column(PickleType)
```
I don't think we need this to be a `PickleType`. The `CloudProvider` class does not really hold any instance attributes and so does not need a persistent representation within the database. Instead, the way to do this is by adding a method to `Archive` with the `sqlalchemy.orm.reconstructor` decorator; this method will be invoked when the object is instantiated from the database. So in this case, we can instantiate the `CloudProvider` instance within such a method.

merrygoat replied:
I think I have managed this.

aplowman commented on hpcflow/api.py:
```diff
@@ -473,7 +474,7 @@ def update_config(name, value, config_dir=None):
 def cloud_connect(provider, config_dir=None):
```
We will need to re-think this function. (I thought I fixed it with the change I made - but obviously that doesn't actually help!)

merrygoat replied:
I cant remember exactly what was wrong with this but I notice that in that function
`provider.check_access()`
the provider object is not instantiated. I have changed this to:
`provider().check_access()`

In testing now it seems to work as expected.

aplowman commented on hpcflow/archive/cloud/cloud.py:
```diff
@@ -1,45 +1,48 @@
-import enum
+class CloudProvider(ABC):
+    """A cloud provider is a service that allows the archival of files to a remote location."""
+    def check_access(self) -> bool:
```
Instead of explicitly raising `NotImplementedError`, we can decorate these ABC methods with `abc.abstractmethod`.

aplowman reviewed (changes requested):
Thanks for this work Peter! The use of an abstract base class makes a lot of sense here.

I think we should add a module-level dict to `hpcflow.archive.cloud.cloud` like this:

```py
CLOUD_PROVIDERS = {
    'dropbox': DropboxCloudProvider,
    'onedrive': OneDriveCloudProvider,
}
```
and do away with the `Archive.cloud_provider_type` attribute. If we use a `@reconstructor`-decorated method in the `Archive` class (see comment), we should then use that method to do something like: `self._cloud_provider = CLOUD_PROVIDERS[self.cloud_provider_name]` where `Archive.cloud_provider_name` is a string column stored in the database that is originally supplied in `Archive.__init__` (e.g. "dropbox").

merrygoat committed:
Change module name for dropbox v11

merrygoat committed:
Move dropbox_cp.py to providers folder

Remove tqdm
Rename _choose_session_type
Add decorators to superclass abstract methods

merrygoat committed:
Remove CloudProviderType object

merrygoat committed:
Use providers dictionary in cloud_connect function

merrygoat commented:
I have tested this latest version on the CSF with a simple workflow with a large file and it seems to work OK. We can discuss more on Friday.

merrygoat committed:
Add retry to dropbox get directory method

merrygoat committed:
Merge branch 'dev' into cloud_archiving

merrygoat committed:
Add retry to Dropbox methods

merrygoat committed:
Fix reconstitution of Archive object

aplowman commented:
Thanks for your work, @merrygoat, sorry I never merged it!
